core/src/main/java/hudson/util/RobustReflectionConverter.java

```diff
@@ -307,9 +308,11 @@ public Object doUnmarshal(final Object result, final HierarchicalStreamReader re
                         implicitCollectionsForCurrentObject = writeValueToImplicitCollection(context, value, implicitCollectionsForCurrentObject, result, fieldName);
                     }
                 }
+            } catch (CriticalXStreamException e) {
+                throw e;
             } catch (XStreamException e) {
                 if (critical) {
-                    throw e;
+                    throw new CriticalXStreamException(e);
```

Comment:
IMO this conversion is confusing. CriticalXXX is supposed to include only a subset of XXX. I also suspect it conflicts with the "graceful" handling described in Javadoc